### PR TITLE
perf(core): add images lazy-loading

### PR DESCRIPTION
**Pull request by Drapegnik**
- https://web.dev/browser-level-image-lazy-loading/

# loaded images
## home
 _ | count | size
-- | -- | --
was | 19 | 3 000,4 KiB
now | 10 | 1 741,4 KiB

## tag/belarus
 _ | count | size
-- | -- | --
was | 42 | 9 113,3 KiB
now | 11 | 2 416,7 KiB

## article/kitaby
 _ | count | size
-- | -- | --
was | 9 | 2 882,2 KiB
now | 3 | 1 051 KiB

### CODE REVIEW

**Comment by Drapegnik** [comment]
[proper images handling](https://trello.com/c/2NnD5uIj/396-proper-images-handling)

File: features/articles/Article.js [hunk 1]
```diff
@@ -64,7 +64,10 @@ const Article = ({
   },
 }) => {
   const { brands, authors, ...tagsObject } = tagsByTopic;
-  const renderTagImage = getTagImageRenderer({ className: b('tag-image') });
+  const renderTagImage = getTagImageRenderer({
```
Comment (UladBohdan) [inline features/articles/Article.js:67]: Is it for those small images of authors and brands?

File: features/layout/blocks/articles/ArticlesByTag2.js [hunk 2]
```diff
@@ -15,7 +15,7 @@ import styles from './articlesByTag2.module.scss';
 
 const b = bem(styles);
 
-const ArticlesByTag2 = ({ block, data }) => {
+const ArticlesByTag2 = ({ block, data, imageLoading }) => {
```
Comment (UladBohdan) [inline features/layout/blocks/articles/ArticlesByTag2.js]: Wouldn't it be more flexible to pass something like "isVisibleInViewport" and then resolve lazy/eager closer to the image?

File: features/layout/blocks/articles/list-block.js [hunk 3]
```diff
@@ -38,44 +38,59 @@ const arrangeListInBlocks = articles =>
...
     const Block = BLOCK_BY_LEVEL[levelName];
-    // eslint-disable-next-line react/no-array-index-key
-    return <Block key={index} articles={block} layout={LAYOUT_BY_LEVEL[levelName]} />;
+    return (
+      <Block
+        // eslint-disable-next-line react/no-array-index-key
+        key={index}
+        articles={block}
+        layout={LAYOUT_BY_LEVEL[levelName]}
+        imageLoading={index === 0 ? imageLoading : 'lazy'}
```
Comment (UladBohdan) [inline features/layout/blocks/articles/list-block.js]: I'd say `index === 1` also should inherit `imageLoading`, probably `index === 2` also.

File: features/layout/blocks/banner/BannerBlock.js [hunk 4]
```diff
@@ -48,6 +48,7 @@ const BannerBlock = ({ block: { banner } }) => {
 
   const linkProps = LINK[banner];
   const Link = linkProps.route ? LocalLink : ExternalLink;
+  const imageLoading = banner === NY2021 ? 'eager' : 'lazy';
```
Comment (UladBohdan) [inline features/layout/blocks/banner/BannerBlock.js]: 1. `ny2021` slug is lower-cased
2. Is it based on current index page layout? Let's not rely on it and pass `imageLoading` into component as you do with other blocks.

File: features/layout/card-blocks-layout.js [hunk 5]
```diff
@@ -12,16 +12,23 @@ import {
...
     return (
-      // eslint-disable-next-line react/no-array-index-key
-      <Block key={index} block={block} data={data} blocks={blocksByType} />
+      <Block
+        // eslint-disable-next-line react/no-array-index-key
+        key={index}
+        block={block}
+        data={data}
+        blocks={blocksByType}
+        imageLoading={index === 0 ? imageLoading : 'lazy'}
```
Comment (UladBohdan) [inline features/layout/card-blocks-layout.js]: Same thoughts as above.

File: features/articles/Article.js (hunk 1 above)
Comment (Drapegnik) [reply features/articles/Article.js:67]: yep

File: features/layout/blocks/articles/list-block.js (hunk 3 above)
Comment (Drapegnik) [reply features/layout/blocks/articles/list-block.js]: no, I checked: first block is enough

File: features/layout/blocks/banner/BannerBlock.js (hunk 4 above)
Comment (Drapegnik) [reply features/layout/blocks/banner/BannerBlock.js]: `NY2021` - is constant

File: features/articles/Article.js (hunk 1 above)
Comment (UladBohdan) [reply features/articles/Article.js:67]: Samples of both text and non-text articles having tags in viewport:

Screen Shot 2020-12-29 at 21 51 05
Screen Shot 2020-12-29 at 21 52 41

So I'd make `inViewport` non-conditional.

File: features/articles/Article.js [hunk 6]
```diff
@@ -96,6 +99,7 @@ const Article = ({
             alt={title}
             sourceSizes={COVER_SIZES}
             baseUrl={images.page}
+            inViewport
```
Comment (UladBohdan) [inline features/articles/Article.js:102]: Why sometimes `inViewport={inViewport}` and sometimes just `inViewport` ?

Comment (UladBohdan) [reply features/articles/Article.js:102]: Oh I got it it's truthful vs passing a param.

File: features/layout/blocks/banner/BannerBlock.js [hunk 7]
```diff
@@ -57,7 +57,8 @@ const BannerBlock = ({ block: { banner } }) => {
           {banner === NY2021 && (
             <div className={cn(b('subtitle'), b(`${banner}-subtitle`))}>{subtitle}</div>
           )}
-          <Picture sources={BANNERS(banner)} alt={title} />
+          {/* FIXME: inViewport */}
```
Comment (UladBohdan) [inline features/layout/blocks/banner/BannerBlock.js]: what's this fixme about?

File: features/layout/card-blocks-layout.js [hunk 8]
```diff
@@ -12,16 +12,25 @@ import {
...
-const CardBlocksLayout = ({ blocks, data }) => {
+const CardBlocksLayout = ({ blocks, data, inViewport }) => {
   const blocksByType = keyBy(blocks, 'type');
   return blocks.map((block, index) => {
     const Block = BLOCKS_BY_TYPE[block.type];
     if (!Block) {
       return null;
     }
+
+    const maxViewportBlockNumber = block.type === 'banner' ? 2 : 1;
```
Comment (UladBohdan) [inline features/layout/card-blocks-layout.js:23]: banners are not the only "short" blocks. row of three is pretty short too. so is diary. 

sample with 3 blocks in viewport:
Screen Shot 2020-12-29 at 22 03 24

let's not make it *that* optimized. let's just make maxViewportBlockNumber constant as 3 or 4.

**Pull request merged**